cmd/createorupdate/update.go

```diff
@@ -0,0 +1,168 @@
+package main
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-06-01/compute"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/Azure/go-autorest/autorest/to"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+	"github.com/openshift/openshift-azure/pkg/log"
+)
+
+func update(cs *api.OpenShiftManagedCluster, p api.Plugin) error {
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		return err
+	}
+
+	ssc := compute.NewVirtualMachineScaleSetsClient(cs.Properties.AzProfile.SubscriptionID)
+	ssc.Authorizer = authorizer
+	vmc := compute.NewVirtualMachineScaleSetVMsClient(cs.Properties.AzProfile.SubscriptionID)
+	vmc.Authorizer = authorizer
+
+	ctx := context.Background()
+
+	err = updateInPlace(ctx, cs, p, ssc, vmc, api.AgentPoolProfileRoleMaster)
+	if err != nil {
+		return err
+	}
+
+	// TODO: updatePlusOne isn't good enough to avoid interruption on our infra
+	// nodes.
+	err = updatePlusOne(ctx, cs, p, ssc, vmc, api.AgentPoolProfileRoleInfra)
+	if err != nil {
+		return err
+	}
+
+	err = updatePlusOne(ctx, cs, p, ssc, vmc, api.AgentPoolProfileRoleCompute)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getCount(cs *api.OpenShiftManagedCluster, role api.AgentPoolProfileRole) int {
+	for _, app := range cs.Properties.AgentPoolProfiles {
+		if app.Role == role {
+			return app.Count
+		}
+	}
+
+	panic("invalid role")
+}
+
+func listVMs(ctx context.Context, cs *api.OpenShiftManagedCluster, vmc compute.VirtualMachineScaleSetVMsClient, role api.AgentPoolProfileRole) ([]compute.VirtualMachineScaleSetVM, error) {
+	vmPages, err := vmc.List(ctx, cs.Properties.AzProfile.ResourceGroup, "ss-"+string(role), "", "", "")
+	if err != nil {
+		return nil, err
+	}
+
+	var vms []compute.VirtualMachineScaleSetVM
+	for vmPages.NotDone() {
+		vms = append(vms, vmPages.Values()...)
+
+		err = vmPages.Next()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return vms, nil
+}
+
+// updatePlusOne creates an extra VM, then runs updateInPlace, then removes the
+// extra VM.
+func updatePlusOne(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin, ssc compute.VirtualMachineScaleSetsClient, vmc compute.VirtualMachineScaleSetVMsClient, role api.AgentPoolProfileRole) error {
+	count := getCount(cs, role)
+
+	log.Infof("setting ss-%s capacity to %d", role, count+1)
+	future, err := ssc.Update(ctx, cs.Properties.AzProfile.ResourceGroup, "ss-"+string(role), compute.VirtualMachineScaleSetUpdate{
+		Sku: &compute.Sku{
+			Capacity: to.Int64Ptr(int64(count) + 1),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	err = future.WaitForCompletion(ctx, ssc.Client)
+	if err != nil {
+		return err
+	}
+
+	// TODO: should wait for the new node to be ready, but I'm not aware of a
```

Comment:
For now, just wait for the node to get to Ready state?

Reply:
@mjudeikis @pweil- for discussion with MSFT.  The difficulties here are:
a) crash-restart at the point where we already added the N+1 node.  In that case, at this point the capacity change is a no-op
b) detecting which VM is the "new VM" (if applicable) and waiting for it to change - don't see a robust way to do this given the available Azure API